Define how versioning works in detail (#103)
ceos-org/ceos-ard: VERSIONING.md

# Versioning <!-- omit in toc -->

The CEOS-ARD Product Family Specifications (PFS) follow [Semantic Versioning 2.0.0](https://semver.org/) (SemVer).
Each PFS is versioned independently as `MAJOR.MINOR.PATCH` (e.g. 5.0.2) and released with a Git tag following the pattern `<PFS>-v<major>.<minor>.<patch>` (see [CONTRIBUTING.md](CONTRIBUTING.md#release-process)).

This document defines what counts as a **major** (i.e. breaking), **minor**, or **patch** change, both for the PFS documents and for the individual building blocks they are compiled from.

- [Guiding Principle](#guiding-principle)
- [How Versions Are Determined](#how-versions-are-determined)
- [Classification by Building Block Type](#classification-by-building-block-type)
  - [Requirements](#requirements)
  - [Glossary](#glossary)
  - [References](#references)
  - [Editorial Content](#editorial-content)
  - [PFS Documents](#pfs-documents)
  - [Templates and Assets](#templates-and-assets)
- [Quick Reference](#quick-reference)
- [Recording Changes](#recording-changes)

## Guiding Principle

> SemVer is defined in terms of a "public API". For CEOS-ARD, the public API of a PFS is its **normative content**: everything a data provider assesses their product against in a self-assessment, and everything a data user relies on when they see a product labelled as CEOS-ARD compliant.

Reviewer: Semantic versioning ([SemVer](https://semver.org/)) is defined in terms of ...

Author: That's already introduced and linked to in line 3:

> The CEOS-ARD Product Family Specifications (PFS) follow [Semantic Versioning 2.0.0](https://semver.org/) (SemVer).

Reviewer: Ah, yes I missed that, sorry.

- **Patch:** The normative meaning does not change at all. Editorial fixes, wording clarifications, formatting, identifier housekeeping, and corrections to non-normative content.

Three consequences of this principle are worth spelling out:

1. **Classify by effect, not by intent or by size.** A one-character typo fix that accidentally changes a limit from `10 m` to `100 m` is a normative change, not a patch. A large restructuring of the whole document that leaves every requirement semantically intact is a patch.
2. **When in doubt, pick the higher level.** It is always safe to release a major version for a non-breaking change; the reverse silently lets products claim compliance with a version whose requirements they may not meet.

> 3. **Product assessments are irrelevant.** Classification depends on the normative effect of the change, not on whether any product assessment currently exists. A change can still be major even if no assessment has been published yet.

Reviewer 2: Rename to: "The existence of prior assessments has no bearing on classification" or similar

Author: Like this?

```suggestion
3. **Product assessments do not affect the classification.** Classification depends only on the normative effect of the change. The existence or non-existence of prior assessments has no bearing on the classification.
```

2. When a PFS is released, its new version number is determined by the **highest level** among all direct or indirect changes that affect that PFS since its last release.

Because building blocks are shared, note:

- A change to a shared building block propagates to **every PFS that uses it**. For example, a single breaking change to `requirements/per-pixel/nodata.yaml` will eventually trigger a new major release of several PFS.
- The `level` recorded on the change describes the change itself. Whether it actually affects a given PFS depends on whether that PFS includes the building block. A new building block that is not (yet) referenced by any `document.yaml` has no version impact at all. Changes to a building block impact the version of a PFS only after the building block has been part of a release of that PFS for the first time.

## Classification by Building Block Type

### Requirements

Requirement files (in folder [`requirements`](./requirements/)) contain the threshold and goal sub-requirements, notes, dependencies, and links to glossary terms and references.

> Threshold requirements are mandatory; goal requirements are optional. Self-assessments also record which goal requirements a product meets, and providers may advertise goal-level compliance.

Reviewer 2: "goal requirements are optional, but provide aspirational guidance"

Author: ```suggestion
Threshold requirements are mandatory; goal requirements are optional, but provide aspirational guidance. Self-assessments also record which goal requirements a product meets, and providers may advertise goal-level compliance.
```

> **Major** - a compliant product may become non-compliant:

Reviewer 2: "a previously compliant product may become non-compliant with a new version of a PFS:"

Author: ```suggestion
**Major** - a previously compliant product may become non-compliant with a new version of a PFS:
```

- Tightening a threshold requirement, e.g.:
  - a numeric limit becomes stricter (geometric accuracy of 5 m instead of 10 m; time to the millisecond instead of the second);
  - a normative verb is strengthened (*should* → *must*, *recommended* → *required*);
  - an additional condition is added ("… and the offset from UTC must be unambiguously identified");
  - a choice is removed ("either X or Y" becomes "X only").

> - Adding a new threshold sub-requirement to an existing requirement file.

Reviewer 2: "e.g., where 'Not required.' was previously the threshold"

Author: ```suggestion
- Adding a new threshold sub-requirement to an existing requirement file, e.g., where 'Not required' was previously the threshold.
```

- Downgrading a threshold requirement to a goal requirement.
- Adding a new goal sub-requirement, tightening an existing goal requirement, relaxing a goal requirement, or removing one.
- Changing a dependency so that it points to a more lenient requirement.

**Patch** - no change to normative meaning:

- Fixing typos, grammar, or punctuation.
- Rewording for clarity without changing meaning (e.g. making the "no threshold requirement applies" wording consistent across requirements).
- Formatting changes.
- Renaming a requirement file or identifier, provided the old identifier is recorded in `history` and all referencing PFS and dependencies are updated.
- Adding, removing, or reordering entries in `glossary` or `references` lists (the linked content itself is classified separately, see below).
- Adding or improving the introductory `description` of a requirement, as long as it only provides context.
- Adding a note that clarifies or documents existing flexibility without changing what is required.
- Filling in `changes` or `history` fields.

Changes to entries in the `metadata` section generally follow the same pattern.

> [!IMPORTANT]

> > Requirement category descriptions must not contain normative requirements. If a requirement-like statement is discovered in a category, move it into a proper requirement file under `requirements` and classify that move by the normative effect of the requirement change. See also the chapter ["Editorial Content"](#editorial-content) for more details.

Reviewer 2: can you please point me to where these category descriptions are / what they are?

Author: The introductory text to each of the (formerly 4) requirement categories: Metadata, Per-Pixel Metadata, etc.

The building blocks for it are here: https://github.com/ceos-org/ceos-ard/tree/main/sections/requirement-categories

Reviewer 2: understood, thanks

### Glossary

> Glossary entries (in the folder [`glossary`](./glossary/)) define terms used in requirements. A definition change can silently change the meaning of every requirement that uses the term, so classify by the **effect on the requirements that use it**:

Reviewer 2: Oh wow I hadn't considered this. What a can of worms...is there some way to provide convenient visibility into which building blocks would be affected by a change in the glossary. This would then need to be cleared with multiple custodians? Tricky..

Author: We currently do not show this. We only show which requirement belongs to which PFS right now. It would indeed be important to show more dependency paths in the Editor. We should open an issue for it in the ceos-ard-server and/or editor (to do).

Reviewer 2: ive added an issue 

https://github.com/ceos-org/ceos-ard/issues/104

Might be in the wrong spot

Author: Yeah, moved it to the Editor

- **Major:** Changing a definition so that a threshold requirement effectively becomes stricter. Example: a threshold requirement demands that clouds are flagged per pixel; redefining *cloud* to include thin cirrus tightens that requirement for every provider whose cloud mask excluded cirrus.
- **Minor:** Changing a definition so that requirements using the term effectively become more lenient, or so that only goal requirements are affected. Example: narrowing the definition of *cloud* so fewer pixels must be flagged.
- **Patch:**
  - Fixing typos or improving the wording of a definition without changing its meaning.
  - Expanding or correcting the long form of an abbreviation (e.g. DEM → Digital Elevation Model).

>   - Adding a new glossary entry (making an already-used term explicit, or preparing a term for future use).

Reviewer 2: This could be a major change depending on the content ?

Author: In theory, yes. It's hard to evaluate this though if no former definition existed and it's unclear how it was meant.

Reviewer 2: id suggest removing the final dot point "Adding a new glossary entry" then

Author: Which leaves this case undefined. I guess it would be better to clarify or extend the current definition, if needed.

- **Patch:**
  - Correcting BibTeX metadata: authors, title, year, DOI, URL, page numbers.
  - Fixing a broken link or pointing to a better-accessible copy of the *same* document.
  - Adding an informative (non-normatively cited) reference.
  - Removing a reference that is no longer cited.
  - Renaming a reference id/file (with the BibTeX entry identifier and all referencing building blocks updated).

### Editorial Content

Everything in [`sections`](./sections/) is editorial and non-normative by default, including:

- introduction chapters (in [`sections/introduction`](./sections/introduction/)),
- requirement categories (in [`sections/requirement-categories`](./sections/requirement-categories/)),
- annexes (in [`sections/annexes`](./sections/annexes/)), and
- other or related narrative content.

Changes to content in `sections` are therefore usually **patch** changes: rewriting explanatory text, updating examples, improving figures, restructuring chapters, or reordering sections.

> [!IMPORTANT]

> > There is one **exception**: if a requirement normatively references specific content in `sections/` (for example, "must follow the procedure in Annex A"), that referenced content is relevant for compliance and must be classified by effect just like requirement text (major/minor).

Reviewer 2: i think we should discourage this kind of cross-referencing. In this example, I would suggest the referenced 'Procedure' probably needs to be in the building block? The cross-referencing is a bit messy.

Author: Yes, fuly agreed. Currently it's done especially in SAR, I think. So is still relevant until we remove such references.

Requirements must remain self-contained in the `requirements` folder (optionally using glossary/reference links). If normative semantics would otherwise depend on text in any editorial content, that normative text must be moved into a requirement and versioned as a requirement change.

### PFS Documents

> The `document.yaml` in each PFS folder ([`pfs/<PFS>`](./pfs/)) selects which building blocks make up a PFS, so list changes here are often the ones that actually change what a provider is assessed against:

Reviewer: The last part of the sentence could be a little clearer; I'm not sure what its adding. A provider is assessed against a (versioned) document but changes to a document could be major/minor/patch, which doesn't necessarily mean 'actually change what a provider is assessed against'.

Author: It refers to additon/removing requirements to a PFS. But I guess we can simply remove it:

```suggestion
The `document.yaml` in each PFS folder ([`pfs/<PFS>`](./pfs/)) selects which building blocks make up a PFS:
```

**Major:**

- Adding a requirement (with a threshold part) to the `requirements` list.
- Replacing a requirement with a stricter variant (e.g. swapping `metadata/time` for a variant that additionally demands per-pixel times at threshold level), including via dependency `replace` overrides.
- Narrowing `applies_to`. Products that fall outside the new scope can no longer be assessed against the new version of this PFS at all.

**Minor:**

- Removing a requirement from the `requirements` list (a relaxation).
- Replacing a requirement with a more lenient variant.
- Adding a requirement that only has goal parts.
- Broadening `applies_to`. Existing products remain in scope and compliant; new kinds of products become eligible.

**Patch:**

> - Moving requirements between categories, reordering requirements or categories, or renaming categories (presentation/numbering only).

Reviewer 2: this would probably have implications on references to sections in metadata? might be a bit confusing as a patch?

Author: References within the document are handled automatically by the tool. Metadata should not rely on order in the PFS. I don't see an issue here.

- Adding, removing, or reordering `introduction` and `annexes` entries.

> - Changes to `title`, `type`, `background`, `authors`, document-level `glossary`/`references` lists, and `changes` entries themselves.

Reviewer 2: would changes to the 'title' and 'type' really be considered patch ? i suppose it doesn't change the requirements, but would be a bit odd

Author: Yeah, that's debatable.
